_posts/2018-03-15-totw-141.md

```diff
@@ -0,0 +1,149 @@
+
+*by Samuel Freilich [(sfreilich@google.com)](mailto:sfreilich@google.com)*
+
+## Two Kinds of Null Pointer Checks
+
+Checking a pointer for null before dereferencing is important to avoid crashes
+and bugs. This can be done in two ways:
+
+```c++
+if (foo) {
+  DoSomething(*foo);
+}
+```
+
+```c++
+if (foo != nullptr) {
+  DoSomething(*foo);
+}
+```
+
+Both of these conditionals have the same semantics given that `foo` is a
+pointer, but the type-checking on the latter is a little tighter. Many types
+in C++ can be implicitly converted to `bool`, and additional caution
+is required when the pointed-to type can itself be converted to `bool`.
+
+Consider the following code which could have two very different meanings:
+
+```c++
+// Bad code
+bool* is_migrated = ...;
+
+// Is this checking that `is_migrated` is not null, or was the actual
+// intent to verify that `*is_migrated` is true?
+if (is_migrated) {
+  ...
+}
+```
+
+This code is clearer:
+
+```c++
+// Good code
+// Looks like a null-pointer check for a bool*
+if (is_migrated != nullptr) {
+  ...
+}
+```
+
+Both styles are acceptable in Google C++ code. So when the underlying type is
+not implicitly convertible to `bool`, follow the style of surrounding code. If
+the value in question is a "smart pointer" like `std::unique_ptr`, the semantics
+and tradeoffs are the same.
+
+## Optional Values and Scoped Assignments
+
+What about (e.g. `absl::optional`) values? They deserve more careful
+consideration.
+
+For example:
+
+```c++
+// Bad code
+absl::optional<bool> b = MaybeBool();
+if (b) { ... }  // What happens when the function returns absl::optional(false)?
+```
+
+This is less clear than the following code:
+
+```c++
+// Better code
+absl::optional<bool> b = MaybeBool();
+if (b.has_value()) { ... }
+```
+
+Note that, in facy, the code snippets above are actually equivalent:
+`absl::optional`'s conversion to `bool` only looks at whether the `optional` is
+full, not at its contents. A reader may find it counterintuitive that
+`optional(false)` is `true`, but it's immediately clear that `optional(false)`
+has a value. Again, it's worth extra caution when the underlying type is
+implicitly convertible to `bool`.
+
+One pattern for optional return-values is to put a variable declaration in the
+conditional of the `if`. This
+[limits the scope](https://google.github.io/styleguide/cppguide.html#Local_Variables)
+of the variable, but involves an implicit conversion to `bool`:
+
+```c++
+if (absl::optional<Foo> foo = MaybeFoo()) {
+  DoSomething(*foo);
+}
+```
+
+However, avoid this pattern if the underlying type is implicitly convertible to
+`bool`.
+
+## "Boolean-like" Enums
+
+Let's say you've taken the advice of [TotW #94](/tips/94) and decided to use an
+`enum` in your function signature instead of a `bool` for better readability at
+the call sites. This kind of refactoring might introduce an implicit conversion
+in the function definition:
+
+```c++
+// Bad code
+void InitGoogle(
```

Comment:
The example in #94 externally is ParseCommandLineFlags.

Reply:
Changed.